contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+    pub pool_size: i64,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+/// The error type for fetching the DatabaseConfig
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingConfig,
+    MissingTable,
+    MissingConnectionUrl,
+    InvalidConnectionUrl,
+    InvalidPoolSize(i64),
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            DatabaseConfigError::MissingConfig => {
+                write!(f, "The active configuration does not contain a `databases` table.")
+            },
+            DatabaseConfigError::MissingTable => {
+                write!(f, "Database-specific configuration was not found for this database")
+            },
+            DatabaseConfigError::MissingConnectionUrl => {
+                write!(f, "The connection URL is missing for this database")
+            },
+            DatabaseConfigError::InvalidConnectionUrl => {
+                write!(f, "The specified connection URL is invalid")
+            },
+            DatabaseConfigError::InvalidPoolSize(invalid_size) => {
+                write!(f, "'{}' is not a valid value for `pool_size`.", invalid_size)
+            },
+        }
+    }
+}
+
+/// This method retrieves the database configuration from the `Rocket.toml` file
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(
+    name: &'a str,
+    from: &'a config::Config
+) -> Result<DatabaseConfig<'a>, DatabaseConfigError> {
+    // Find the first `databases` config that's a table with a key of 'name'
+    // equal to `name`.
+    let connection_config = from.get_slice("databases")
+        .map_err(|_| DatabaseConfigError::MissingConfig)?
+        .iter()
+        .filter_map(|config| config.as_table())
+        .filter(|table| table.get("name").map(|n| n.as_str() == Some(name)).unwrap_or(false))
+        .next()
+        .ok_or(DatabaseConfigError::MissingTable)?;
+
+    let maybe_url = connection_config.get("url")
+        .ok_or(DatabaseConfigError::MissingConnectionUrl)?;
+
+    let url = maybe_url.as_str().ok_or(DatabaseConfigError::InvalidConnectionUrl)?;
+
+    let pool_size = connection_config.get("pool_size")
+        .and_then(Value::as_integer)
+        .unwrap_or(from.workers as i64);
+
+    if pool_size < 1 {
+        return Err(DatabaseConfigError::InvalidPoolSize(pool_size));
+    }
+
+    Ok(DatabaseConfig { url, pool_size, extras: connection_config })
```

Comment:
It does just contain the entire config. It shouldn't be too difficult to strip out what's already contained in the `DatabaseConfig` type and only pack in the true "extras." Additionally, this will be helpful in implementing TLS support for MySQL and Postgres.

Reply:
I don't think it's necessarily a problem; I just wanted to point out the inconsistency so it's not a surprise later.

Reply:
Sure. If it is an inconsistency, I'd rather resolve it. It also would make it easier to use the `extras` because it would only be what is extra and what hasn't already been placed in the `DatabaseConfig` object.